File: packages/design-ir/src/migrations.ts
import { canonicalize } from "./canonical";
import { IrRuntimeError, type DesignDocument, type MigrationStep } from "./types";
import { parseDocument } from "./validation";

function recordMigration(
  document: DesignDocument,
  from: string,
  to: string,
): DesignDocument {
  const provenance = document.metadata.migration_provenance;
  const current = Array.isArray(provenance) ? provenance : [];
  return {
    ...structuredClone(document),
    schema_version: to,
    metadata: {
      ...structuredClone(document.metadata),
      migration_provenance: [
        ...current,
        {
          from,
          to,
          source_canonical: canonicalize(document),
        },
      ],
    },
  };
}

export const MIGRATIONS: readonly MigrationStep[] = [
  {
    from: "1.0",
    to: "1.1",
    migrate: (document) => recordMigration(document, "1.0", "1.1"),
  },
  {
    from: "1.1",
    to: "2.0",
    migrate: (document) => recordMigration(document, "1.1", "2.0"),
  },
];

export function migrate(document: DesignDocument, targetVersion: string): DesignDocument {
  let current = parseDocument(document);
  const seen = new Set<string>();
  while (current.schema_version !== targetVersion) {
    if (seen.has(current.schema_version)) {
      throw new IrRuntimeError({
        code: "IR_VERSION_UNSUPPORTED",
        message: `migration cycle from ${current.schema_version}`,
      });
    }
    seen.add(current.schema_version);
    const step = MIGRATIONS.find((candidate) => candidate.from === current.schema_version);
    if (!step) {
      throw new IrRuntimeError({
        code: "IR_VERSION_UNSUPPORTED",
        message: `no migration path from ${current.schema_version} to ${targetVersion}`,
      });
    }
    current = parseDocument(step.migrate(current));
  }
  return current;
}

File: packages-py/design_ir/migrations.py
from __future__ import annotations

from copy import deepcopy

from .canonical import canonicalize
from .models import DesignDocument, IrIssue, IrRuntimeError
from .validate import parse_document


def _record(document: DesignDocument, source: str, target: str) -> DesignDocument:
    result = deepcopy(document)
    provenance = result.setdefault("metadata", {}).get("migration_provenance")
    history = list(provenance) if isinstance(provenance, list) else []
    history.append(
        {
            "from": source,
            "to": target,
            "source_canonical": canonicalize(document),
        }
    )
    result["metadata"]["migration_provenance"] = history
    result["schema_version"] = target
    return result


MIGRATIONS = {
    "1.0": ("1.1", lambda document: _record(document, "1.0", "1.1")),
    "1.1": ("2.0", lambda document: _record(document, "1.1", "2.0")),
}


def migrate(document: DesignDocument, target_version: str) -> DesignDocument:
    current = parse_document(document)
    seen: set[str] = set()
    while current["schema_version"] != target_version:
        source = current["schema_version"]
        if source in seen or source not in MIGRATIONS:
            raise IrRuntimeError(
                IrIssue(
                    "IR_VERSION_UNSUPPORTED",
                    f"no migration path from {source} to {target_version}",
                )
            )
        seen.add(source)
        _, step = MIGRATIONS[source]
        current = parse_document(step(current))
    return current
